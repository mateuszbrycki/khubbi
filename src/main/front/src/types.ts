import {DateTimeFormatter, LocalDateTime, ZonedDateTime, ZoneId} from "js-joda";
import {v4 as uuidv4} from 'uuid';


export class EventDate {
    value: ZonedDateTime

    public constructor(value: ZonedDateTime) {
        this.value = value
    }

    public toString(): string {
        return this.value.toString();
    }

    public format(format: string): string {
        return this.value.format(DateTimeFormatter.ofPattern(format))
    }

    public toISODateTime(): string {
        return this.value.format(DateTimeFormatter.ISO_OFFSET_DATE_TIME);
    }

    public atUTC(): EventDate {
        return new EventDate(this.value.withZoneSameLocal(ZoneId.UTC))
    }

    public static now(): EventDate {
        return new EventDate(ZonedDateTime.now());
    }

    public static ofDateAndTime(date: string): EventDate {
        return new EventDate(LocalDateTime.parse(date)
            .atZone(ZoneId.SYSTEM));
    }

    public static ofZoneDateAndTime(date: string): EventDate {
        return new EventDate(ZonedDateTime.parse(date));
    }

}

export class Photo {
    readonly date: EventDate;
    readonly id: string;
    readonly description: string;
    readonly photo: string;

    constructor(id: string, date: EventDate, description: string, photo: string) {
        this.date = date;
        this.id = id;
        this.description = description;
        this.photo = photo;
    }
}

export class Note {
    readonly date: EventDate;
    readonly id: string;
    readonly note: string;

    constructor(id: string, date: EventDate, note: string) {
        this.date = date;
        this.id = id;
        this.note = note;
    }
}

export type Event = Photo | Note

export interface User {
    readonly id: string;
    readonly email: string;
    readonly createdAt: string;
}

export class EventForms {
    public static readonly NOTE = new EventForms('Note');
    public static readonly PICTURE = new EventForms('Picture');
    public static readonly EVENT = new EventForms('Event');
    public static readonly EXPENSE = new EventForms('Expense');
    public static readonly REMINDER = new EventForms('Reminder');

    private constructor(public readonly name: string) {
    }
}

export class AlertType {

    public static readonly ERROR = new AlertType('ERROR', "Error", "danger");
    public static readonly WARNING = new AlertType('WARNING', "Warning", "warning");
    public static readonly SUCCESS = new AlertType('SUCCESS', "Success", "success");

    private constructor(public readonly type: string, public readonly title: string, public readonly bg: string) {
    }
}

export class AlertMessage {
    public readonly id: string;
    public readonly message: string;
    public readonly type: AlertType;

    constructor(id: string, message: string, type: AlertType) {
        this.id = id
        this.message = message
        this.type = type
    }

    public static error(message: string) {
        return new AlertMessage(uuidv4(), message, AlertType.ERROR)
    }

    public static success(message: string) {
        return new AlertMessage(uuidv4(), message, AlertType.SUCCESS)
    }

    public static warning(message: string) {
        return new AlertMessage(uuidv4(), message, AlertType.WARNING)
    }
}