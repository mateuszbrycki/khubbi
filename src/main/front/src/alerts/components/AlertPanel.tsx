import React from "react";
import {List} from "immutable";
import {AlertMessage} from "../../types";
import {Toast, ToastContainer} from "react-bootstrap";

export interface AlertPanelProps {
    readonly alerts: List<AlertMessage>,
    readonly alertTimeoutMs: number
}

export interface AlertPanelActionProps {
    readonly hideAlert: (alertMessage: AlertMessage) => void
}

export const AlertPanel: React.FC<AlertPanelProps & AlertPanelActionProps> = (props) => {

    const {alerts, alertTimeoutMs, hideAlert} = props

    const alertToasts = alerts.map((alert: AlertMessage) =>
        <Toast key={alert.id}
               id={alert.id.toString()}
               show={true}
               bg={alert.type.bg}
               onClose={() => {
                   hideAlert(alert)
               }} delay={alertTimeoutMs} autohide>
            <Toast.Header>
                <strong className="me-auto">{alert.type.title}</strong>
            </Toast.Header>
            <Toast.Body>
                {alert.message}
            </Toast.Body>
        </Toast>)

    return <>
        {alerts.size <= 0 ? <></> :
            <ToastContainer position="bottom-end" className="p-3" style={{zIndex: 1}}>
                {alertToasts}
            </ToastContainer>
        }
    </>

}