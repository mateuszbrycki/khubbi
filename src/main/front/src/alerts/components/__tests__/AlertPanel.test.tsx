import ShallowRenderer from "react-test-renderer/shallow";
import {AlertPanel} from "../AlertPanel";
import {List} from "immutable";
import {AlertMessage, AlertType} from "../../../types";
import {render} from "@testing-library/react";
import {setTimeout} from 'timers/promises';

describe('Alert Panel', () => {
    const renderer = ShallowRenderer.createRenderer()
    const alerts = List.of(new AlertMessage("id-1", "test-error", AlertType.ERROR),
        new AlertMessage("id-2", "test-success", AlertType.SUCCESS),
        new AlertMessage("id-3", "test-warning", AlertType.WARNING),
    )

    it('renders empty list when no alerts', () => {
        renderer.render(<AlertPanel alerts={List.of()} hideAlert={() => null} alertTimeoutMs={1}/>)
        expect(renderer.getRenderOutput()).toMatchSnapshot()
    })

    it('renders alerts', () => {
        renderer.render(<AlertPanel alerts={alerts} hideAlert={() => null} alertTimeoutMs={1}/>)
        expect(renderer.getRenderOutput()).toMatchSnapshot()
    })

    it('calls hide alert when closing the toast', async () => {
        const hideAlertMock = jest.fn(() => Promise.resolve());
        render(<AlertPanel alerts={alerts} hideAlert={hideAlertMock} alertTimeoutMs={1}/>)

        await setTimeout(2);

        expect(hideAlertMock).toHaveBeenCalled()

    })

})