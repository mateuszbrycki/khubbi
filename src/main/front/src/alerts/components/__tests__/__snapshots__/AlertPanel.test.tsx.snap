// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alert Panel renders alerts 1`] = `
<React.Fragment>
  <ToastContainer
    className="p-3"
    position="bottom-end"
    style={
      {
        "zIndex": 1,
      }
    }
  >
    Immutable.List [
      <Toast
        autohide={true}
        bg="danger"
        delay={1}
        id="id-1"
        onClose={[Function]}
        show={true}
      >
        <ToastHeader>
          <strong
            className="me-auto"
          >
            Error
          </strong>
        </ToastHeader>
        <ToastBody>
          test-error
        </ToastBody>
      </Toast>,
      <Toast
        autohide={true}
        bg="success"
        delay={1}
        id="id-2"
        onClose={[Function]}
        show={true}
      >
        <ToastHeader>
          <strong
            className="me-auto"
          >
            Success
          </strong>
        </ToastHeader>
        <ToastBody>
          test-success
        </ToastBody>
      </Toast>,
      <Toast
        autohide={true}
        bg="warning"
        delay={1}
        id="id-3"
        onClose={[Function]}
        show={true}
      >
        <ToastHeader>
          <strong
            className="me-auto"
          >
            Warning
          </strong>
        </ToastHeader>
        <ToastBody>
          test-warning
        </ToastBody>
      </Toast>,
    ]
  </ToastContainer>
</React.Fragment>
`;

exports[`Alert Panel renders empty list when no alerts 1`] = `
<React.Fragment>
  <React.Fragment />
</React.Fragment>
`;
